services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/UploadDirectoryConfiguration.java

```diff
@@ -0,0 +1,183 @@
+ * use the SDK default values.
+ *
+ * <p>Use {@link #builder()} to create a set of options.</p>
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class UploadDirectoryConfiguration implements ToCopyableBuilder<UploadDirectoryConfiguration.Builder,
+    UploadDirectoryConfiguration> {
+
+    private final Boolean followSymbolicLinks;
+    private final Integer maxDepth;
+    private final Boolean recursive;
+
+    public UploadDirectoryConfiguration(DefaultBuilder builder) {
+        this.followSymbolicLinks = builder.followSymbolicLinks;
+        this.maxDepth = builder.maxDepth;
+        this.recursive = builder.recursive;
+    }
+
+    /**
+     * @return whether to follow symbolic links
+     */
+    public Optional<Boolean> followSymbolicLinks() {
+        return Optional.ofNullable(followSymbolicLinks);
+    }
+
+    /**
+     * @return the maximum number of directory levels to traverse
+     */
+    public Optional<Integer> maxDepth() {
+        return Optional.ofNullable(maxDepth);
+    }
+
+    /**
+     * @return whether to recursively upload all files under the specified directory
+     */
+    public Optional<Boolean> recursive() {
+        return Optional.ofNullable(recursive);
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return new DefaultBuilder(this);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        UploadDirectoryConfiguration that = (UploadDirectoryConfiguration) o;
+
+        if (!Objects.equals(followSymbolicLinks, that.followSymbolicLinks)) {
+            return false;
+        }
+        if (!Objects.equals(maxDepth, that.maxDepth)) {
+            return false;
+        }
+        return Objects.equals(recursive, that.recursive);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = followSymbolicLinks != null ? followSymbolicLinks.hashCode() : 0;
+        result = 31 * result + (maxDepth != null ? maxDepth.hashCode() : 0);
+        result = 31 * result + (recursive != null ? recursive.hashCode() : 0);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("UploadDirectoryConfiguration")
+                       .add("followSymbolicLinks", followSymbolicLinks)
+                       .add("maxDepth", maxDepth)
+                       .add("recursive", recursive)
+                       .build();
+    }
+
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+    public interface Builder extends CopyableBuilder<Builder, UploadDirectoryConfiguration> {
+
+        /**
+         * Specify whether to recursively upload all files under the specified directory
+         *
+         * @param recursive whether enable recursive upload
+         * @return This builder for method chaining.
+         */
```

Comment:
"Default to true".

Are there any edge-cases we should document? E.g. "Links are excluded/not followed... see followSymbolicLinks", "The current user must have read/execute access to all directories and read access to all files", "hidden files are included/excluded", "wildcards like * / ** are treated literally and not expanded" (I guess this one is better on the directory parameter, but you get the point), etc.

I really think we should raise the bar on our javadoc in our high level libraries to try to anticipate most of what our customer might encounter. Us documenting something once can save hundreds of people from needing to figure it out on their own.

Reply:
+1 to raising the bar on our Javadoc. I got overwhelmed when writing boilerplate code(builders are so verbose!), so the Javadoc is a bit sloppy 😅. Will fix it.